smartsim/_core/launcher/step/step.py

```diff
@@ -87,10 +93,11 @@ def get_colocated_launch_script(self) -> str:
         write_colocated_launch_script(script_path, db_log_file, db_settings)
         return script_path
 
+    # pylint: disable=no-self-use
     def add_to_batch(self, step: Step) -> None:
         """Add a job step to this batch
 
         :param step: a job step instance e.g. SrunStep
         :type step: Step
         """
-        raise NotImplementedError
+        raise SmartSimError("add_to_batch not implemented for this step type")
```

Comment:
Why the change of exception types? This seems like a bad idea

Reply:
NotImplementedError makes the linter treat this class as abstract and it has some downstream effects. Since the non-typed code was loose, it results in either implementing an unnecessary method on Jsrun/Aprun/Local/etc Steps or swapping this exception type out.

I chose the exception route:


```
❯ make check-lint
************* Module smartsim._core.launcher.step.lsfStep
smartsim/_core/launcher/step/lsfStep.py:114:0: W0223: Method 'add_to_batch' is abstract in class 'Step' but is not overridden in child class 'JsrunStep' (abstract-method)
************* Module smartsim._core.launcher.step.alpsStep
smartsim/_core/launcher/step/alpsStep.py:40:0: W0223: Method 'add_to_batch' is abstract in class 'Step' but is not overridden in child class 'AprunStep' (abstract-method)
************* Module smartsim._core.launcher.step.localStep
smartsim/_core/launcher/step/localStep.py:35:0: W0223: Method 'add_to_batch' is abstract in class 'Step' but is not overridden in child class 'LocalStep' (abstract-method)
************* Module smartsim._core.launcher.step.step
smartsim/_core/launcher/step/step.py:33:0: W0611: Unused SmartSimError imported from smartsim.error.errors (unused-import)
************* Module smartsim._core.launcher.step.slurmStep
smartsim/_core/launcher/step/slurmStep.py:109:0: W0223: Method 'add_to_batch' is abstract in class 'Step' but is not overridden in child class 'SrunStep' (abstract-method)
************* Module smartsim._core.launcher.step.mpiStep
smartsim/_core/launcher/step/mpiStep.py:41:0: W0223: Method 'add_to_batch' is abstract in class 'Step' but is not overridden in child class '_BaseMPIStep' (abstract-method)
smartsim/_core/launcher/step/mpiStep.py:149:0: W0223: Method 'add_to_batch' is abstract in class 'Step' but is not overridden in child class 'MpiexecStep' (abstract-method)
smartsim/_core/launcher/step/mpiStep.py:167:0: W0223: Method 'add_to_batch' is abstract in class 'Step' but is not overridden in child class 'MpirunStep' (abstract-method)
smartsim/_core/launcher/step/mpiStep.py:185:0: W0223: Method 'add_to_batch' is abstract in class 'Step' but is not overridden in child class 'OrterunStep' (abstract-method)
```

Reply:
I created a follow-up ticket for this item to be reviewed further